src/rollup/index.ts

```diff
@@ -124,357 +103,389 @@ function getInputOptions(rawInputOptions: GenericConfigObject): any {
+				applyBuildEndHook(graph);
 				timeEnd('BUILD', 1);
 
+				// TODO: deprecate legacy single chunk return
+				let singleInputChunk: Chunk;
+				let imports: string[], exports: string[];
+				if (!inputOptions.experimentalPreserveModules) {
+					if (
+						typeof inputOptions.input === 'string' ||
+						(inputOptions.input instanceof Array && inputOptions.input.length === 1)
+					) {
+						singleInputChunk = chunks.find(chunk => chunk.entryModule !== undefined);
+						imports = singleInputChunk.getImportIds();
+						exports = singleInputChunk.getExportNames();
+					}
+				}
+
 				// ensure we only do one optimization pass per build
 				let optimized = false;
 
-				function generate(rawOutputOptions: GenericConfigObject) {
+				function generate(rawOutputOptions: GenericConfigObject, isWrite: boolean) {
 					const outputOptions = normalizeOutputOptions(inputOptions, rawOutputOptions);
 
-					if (typeof outputOptions.file === 'string')
-						error({
-							code: 'INVALID_OPTION',
-							message: 'When code-splitting, the "dir" output option must be used, not "file".'
-						});
-
-					if (outputOptions.format === 'umd' || outputOptions.format === 'iife') {
-						error({
-							code: 'INVALID_OPTION',
-							message:
-								'UMD and IIFE output formats are not supported with the experimentalCodeSplitting option.'
-						});
+					if (inputOptions.experimentalCodeSplitting) {
+						if (typeof outputOptions.file === 'string' && typeof outputOptions.dir === 'string')
+							error({
+								code: 'INVALID_OPTION',
+								message:
+									'Build must set either output.file for a single-file build or output.dir when generating multiple chunks.'
+							});
+						if (chunks.length > 1) {
+							if (outputOptions.format === 'umd' || outputOptions.format === 'iife')
+								error({
+									code: 'INVALID_OPTION',
+									message:
+										'UMD and IIFE output formats are not supported with the experimentalCodeSplitting option.'
+								});
+
+							if (outputOptions.sourcemapFile)
+								error({
+									code: 'INVALID_OPTION',
+									message: '"sourcemapFile" is only supported for single-file builds.'
+								});
+						}
+						if (!singleInputChunk && typeof outputOptions.file === 'string')
+							error({
+								code: 'INVALID_OPTION',
+								message:
+									'When providing multiple entry points, the output.dir option must be used, not output.file.'
+							});
 					}
 
-					if (outputOptions.sourcemapFile)
-						error({
-							code: 'INVALID_OPTION',
-							message: '"sourcemapFile" is only supported for single-file builds.'
-						});
+					if (!outputOptions.file && inputOptions.experimentalCodeSplitting)
+						singleInputChunk = undefined;
 
 					timeStart('GENERATE', 1);
 
-					const generated: { [chunkName: string]: OutputChunk } = {};
+					// populate asset files into output
+					const assetFileNames = outputOptions.assetFileNames || 'assets/[name]-[hash][ext]';
```

Comment:
It would be really nice if "[ext]"  would not contain the "." as in this case, I could use a pattern such as "[ext]/[name].[ext]" i.e. grouping different types of assets into different folders.

Reply:
The issue with this is that in theory we support assets without an extension, so the `[ext]` pattern can be the empty string. Perhaps we could introduce a new variable - `[extname]` or something like that?

Reply:
Note we can always move to supporting a function for these patterns as well, with a `getVar` function argument itself - 

```js
assetFileNames: (vars) => {
  if (vars.ext === '.css')
    return `assets/style/${vars.name}`;
  return `assets/${vars.name}`;
}
```

Reply:
Note that this is the same output as the NodeJS `path.extname`, by inluding the dot.

Perhaps instead of `[ext]` it should be called `[extname]` with `[ext]` being the version without the dot? That might make it more obvious to Node users too.

Reply:
Sounds good